### ISSUE TITLE
Able to limit the rows returned after other filters

### ISSUE THREAD

**Issue by qc00** [man-group/ArcticDB#245]
Currently, the `row_range/head/tail()` filter is applied before any other filter and conflicts with the `date_range` filter. Most other databases support a `top(n)` or `rownum < n` feature and there's [user demand](https://github.com/man-group/ArcticDB/issues/243#issuecomment-1515133551).

We have to be careful to avoid confusion with the existing `row_range/head/tail()`.

Ideally, we add this to the evolved query builder API so the ordering is clear.

**Comment by sharp2448** [comment]
Nice, I would love to see that feature also.

**Comment by bulltools** [comment]
This feature would indeed add flexibility and be helpful in a number of use cases. Appreciate you raising this issue. Have a great weekend!

**Comment by aschonfeld** [comment]
Same! Please add this feature

**Comment by bulltools** [comment]
I don't mean to be annoying, I know you guys are working hard, but are you able to share an update or ETA for this feature?

**Comment by notamicrodose** [comment]
id like to see this feature added as well ^^

**Comment by Klaim** [comment]
This was discussed privately but I'm moving the discussion here for aync public work.

1. We discussed this issue with @jjerphan and we think that it is a bit underspecified, it needs some clarifications like use cases that should work with the change. We both believe having understood the idea (and Alex clarified most of it in a private meeting) but we have doubts about some details. Providing some examples would be helpful to avoid trying wrong solutions again (there is a similar situation in #976) .

2. Julien and I think that to solve this kind of issue a separation of the request representation and the DAG of tasks to execute to treat the request properly is necessary. The DAG would be built by a query-planner/compositor/whatever-you-name-it which reads the request as value (a sequence of clauses , no processing code in clauses) and output the task DAG to execute from it. That way fine-grained ordering of operations is simple and decorelated from how the request looks like to users.
(the recent switch to an ECS-like system to handle the processing seems to confirm this idea as the task DAG would be the Systems in ECS here).

If we are correct, that is if we are right that we understand the feature correctly and iff there is no long-term-viable shortcut to achieve this feature properly without separating clauses and the task dag,
then this seems like a quite important architectural change and personally I dont feel like my understanding of the overall design of arcticdb's internals is enough to do this kind of change at the moment. It would also affect the whole processing system and several tasks listed in [this epic](https://github.com/man-group/ArcticDB/issues/956).

@alexowens90 Could you help with point 1 and maybe confirm if point 2 seems correct to you?

Also thanks to @alexowens90 @jjerphan @Hind-M and @JohanMabille for helping out on the different aspects of this task.

**Comment by willdealtry** [comment]
The way it works at the moment is that the Clause classes (which are perhaps misnamed), are a sequence of stateless behaviours that operate independently on an entity, which might be a row-group, a column group, an individual segment or the all of the segments from a particular stage (the latter in operations that for whatever reason cannot be parallelized).

The DAG of processing stages is currently just a list (actually a std::vector), although it's possible it will evolve into something more complex in the future - probably at the point where we implement joins there will be processing stages that can be executed in parallel. Note that this is not the case for expressions (as in the expression within a ProjectClause, which create a new column on-the-fly), these pull the data up from the left and right expression nodes (in the case of a binary operator) and will use all available cores. For clauses we might always keep things linear as we want to avoid the situation where we read in a lot of data that there aren't cores available to do anything with, and it's easier to apply back-pressure in a simpler DAG.

I think that in this specific case it should be possible to implement it without far-reaching changes. The only tricky bit is that if a head or tail follows (for example) a filter, the slice information will be correct about the number of rows a segment contains, but not necessarily about its offset. As a concrete example, if we have two row-groups with one segment each (i.e. no column slicing), each with 100 rows, a filter clause might reduce them to 20 rows each. The filter clause will adjust the row-range in the slice, but to adjust the offset, the thread executing the filter on rows 100-200 would need to know the result of the filter on rows 0-100, creating a data dependency that we want to avoid. This is particularly important because in the case where we have multiple clauses to execute on a segment that are all independent, for example a filter followed by a column projection, we actually only schedule a single task which does both things one after another while (hopefully) much of the data is in cache. The upshot of this is that the row-ranges will be 0-20 and 100-120, instead of 0-20 and 20-40, which is what they really are. Obviously this needs fixing up at the end of the processing pipeline in order for us to produce a coherent result set, but we try to postpone that work as much as possible.

In the case of a head or tail, the clause should indicate to the pipeline that it needs a barrier (i.e. it needs to wait for all preceding computation to finish) by returning true from the method requires_repartition. This will cause the pipeline to do a collect on all the preceding data, and pass the list of Composite<EntityID> to the repartition method on the clause.(N.B. the composite part is going to go away, in favour of a single entity per processing unit with as many components as required).

This repartition method needs to scan through the list of results from the previous computation, either from the beginning in the case of head or the end in the case of tail, accumulating segments and reading the row information from the slices until it has enough rows to satisfy the head or tail parameter. It can then discard the rest, and pass the required segments into the body of the clause where they can be truncated and coalesced to produce the correct output. The code to truncate a segment and also to take a group of small segments and combine them into a larger one is already available, so the problem is mostly one of getting the wiring correct. 

In terms of testing it would make sense to concentrate on head and tail following a filter, as filters both alter the row-count of a segment and preserve  column slicing, so they should expose most if not all of the edge-cases that you're likely to encounter. There is a series of tests written by @alexowens90 that checks that clauses can be combined with each other, once things are looking good it would be an idea to add some basic tests of head and tail following aggregation and column projection as well.

Let me know if there are any questions.

**Comment by alexowens90** [comment]
The most simple use case we would like to support:
```
# Use tiny segment so there are 2 rows per segment
def test_head_after_filter(lmdb_version_store_tiny_segment):
    lib = lmdb_version_store_tiny_segment
    df = pd.DataFrame({"col": [0, 1, 0, 1, 0, 1, 0, 1, 0, 1]}, index=pd.date_range("2023-01-01", periods=10))
    lib.write("sym", df)
    q = QueryBuilder()
    q = q[q["col"] == 0]
    q = q.head(2)
    expected = df[df["col"] == 0].head(2)
    received = lib.read("sym", query_builder=q).data
    assert_frame_equal(expected, received)
>>> expected
                    col
2023-01-01    0
2023-01-03    0
```
i.e. perform a filter as normal, and then only keep the first (or last for `tail`) `n` rows of the result.

Another use case we would like to support (possibly in another PR) is to allow `head/tail` after a `date_range` filter. Using the `QueryBuilder`, this would look like:
```
q = q.date_range(pd.Timestamp("2023-01-03"), pd.Timestamp("2023-01-06")).head(1)
```
i.e. retain only the first `n` rows in the specified date range. There are specialised `head` and `tail` methods on the `Library` class, so we could also make these work with `date_range` parameters similar to `read`. When both `date_range` and `query_builder` parameters are provided to `read`, we convert the `date_range` filter into a clause and prepend this to the provided clauses (there is a zero-copy code path when `query_builder` is not provided that we need to keep).
